### DISCUSSION TITLE
How to make it search globally?

### DISCUSSION THREAD

**Discussion by eunjae-lee** [johnlindquist/kit#157] [Q&A]
Hi, I just installed ScriptKit and it looks great.

And I'm wondering if I can search globally across all the scripts without choosing one.
I didn't find any way to do so.

For example, after opening it, I need to choose "App Launcher" and then start typing the app name.
However what if I can just type the app name in the first place? It will for sure trigger all the scripts, which can be a potential performance issue? But if we can opt-in with this behavior and find a way to optimize it, then it could replace Alfred on my end.

What do you think?

**Comment by johnlindquist** [comment]
I would likw to understand your scenario a bit better. Search apps and what else?

Many scripts don't "search" and some scripts generate results like searching google.

We can definitely combine results of multiple scripts into a single script and give that it's own keyboard shortcut, but I need to understand the end goal better.

**Reply by eunjae-lee** [reply]
Hey, I'm happy to provide my usecases in mind.

- Type `cmd + ;` and type `Co` and it shows apps with name including `Co`.
- Type `cmd + ;` and type `#998877` and it shows the results of color calculation.
- Type `cmd + ;` and type `Meeting note` and it searches for a snippet manager script and shows the filtered snippets (and if I hit enter it will copy it to clipboard or type the string, or whatever).
- Type `cmd + ;` and type `fs readFile` and it will show search results from Node.js documentation.

For now, whatever I want to do, I need to choose "a script" first and then perform the action. This is flattening the workflow one level. Does it make sense to you?

**Reply by johnlindquist** [reply]
Yes, the idea of flattening based on user input makes sense. These behaviors are achieved by writing a script that switches behaviors based on the first few characters of input:

Here's a pseudo example:

```js
let colors = [
  {
    name: "Red",
    value: {
      script: "my-color-script",
      value: "#ff0000",
    },
  },
  {
    name: "Green",
    value: {
      script: "my-color-script",
      value: "#00ff00",
    },
  },
  {
    name: "Blue",
    value: {
      script: "my-color-script",
      value: "#0000ff",
    },
  },
]

let nodeDocs = [
  {
    name: "readFile",
    value: {
      script: "open-node-docs",
      value: "readFile",
    },
  },
  {
    name: "writeFile",
    value: {
      script: "open-node-docs",
      value: "writeFile",
    },
  },
]

let apps = [
  {
    name: "ableton",
    value: {
      script: "app-launcher",
      value: "ableton",
    },
  },
  {
    name: "chrome",
    value: {
      script: "app-launcher",
      value: "app-launcher",
    },
  },
  {
    name: "code",
    value: {
      script: "app-launcher",
      value: "code",
    },
  },
]

let { value, script } = await arg(
  `Type something`,
  input => {
    if (input.startsWith("#")) {
      let colorName = input.replace(/^#\s*/, "")
      return colors.filter(({ name }) =>
        name.toLowerCase().startsWith(colorName)
      )
    }

    if (input.startsWith("node")) {
      let docsName = input.replace(/^node\s*/, "")
      return nodeDocs.filter(({ name }) =>
        name.startsWith(docsName)
      )
    }

    //default
    return apps
  }
)

await run(script, value)
```

I'll look into ways of making this easier in future versions or providing scripts that do this based on common behaviors.

**Reply by eunjae-lee** [reply]
It's a meta script. It's a good workaround.
Thanks!

**Reply by johnlindquist** [reply]
I'm also planning on `//Shortcode: g` metadata. Typing the shortcode then space in the menu will run the script. This will be much easier than putting together meta scripts. :)

**Reply by eunjae-lee** [reply]
Yes! That sounds much better :)
